Team417_CENTERSTAGE/src/main/java/org/firstinspires/ftc/team417_CENTERSTAGE/apriltags/AprilTagPoseEstimator.java

```diff
@@ -0,0 +1,328 @@
+package org.firstinspires.ftc.team417_CENTERSTAGE.apriltags;
+
+import static org.firstinspires.ftc.team417_CENTERSTAGE.roadrunner.MecanumDrive.isDevBot;
+
+import android.util.Size;
+
+import com.acmerobotics.dashboard.config.Config;
+import com.acmerobotics.roadrunner.Pose2d;
+import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
+import com.qualcomm.robotcore.hardware.DigitalChannel;
+import com.qualcomm.robotcore.hardware.HardwareMap;
+
+import org.firstinspires.ftc.robotcore.external.hardware.camera.BuiltinCameraDirection;
+import org.firstinspires.ftc.robotcore.external.hardware.camera.WebcamName;
+import org.firstinspires.ftc.team417_CENTERSTAGE.baseprograms.BaseOpMode;
+import org.firstinspires.ftc.team417_CENTERSTAGE.utilityclasses.AprilTagInfo;
+import org.firstinspires.ftc.team417_CENTERSTAGE.utilityclasses.InfoWithDetection;
+import org.firstinspires.ftc.team417_CENTERSTAGE.utilityclasses.Pose;
+import org.firstinspires.ftc.vision.VisionPortal;
+import org.firstinspires.ftc.vision.apriltag.AprilTagDetection;
+import org.firstinspires.ftc.vision.apriltag.AprilTagProcessor;
+
+import java.util.ArrayList;
+
+@Config
+public class AprilTagPoseEstimator {
+    public static double CAMERA_LATENCY = 640; // latency between april tag shown and detection in ms
+
+    public static final boolean USE_WEBCAM = true;  // true for webcam, false for phone camera
+
+    public HardwareMap myHardwareMap;   // gain access to camera in hardwareMap
+
+    public LinearOpMode myOpMode;   // gain access to methods in the calling OpMode.
+
+    /**
+     * The variable to store our instance of the AprilTag processor.
+     */
+    public AprilTagProcessor aprilTag;
+
+    /**
+     * The variable to store our instance of the status light.
+     */
+    public DigitalChannel statusLight;
+
+    /**
+     * The variable to store our instance of the vision portal.
+     */
+    public VisionPortal visionPortal;
+
+    // For this concept: supposed location of the robot
+    Pose robotPoseEstimate = new Pose(0, 0, 0);
+
+    // Define a constructor that allows the OpMode to pass a reference to itself
+    //   in order to allow this class to access the camera hardware
+    public AprilTagPoseEstimator(LinearOpMode opmode) {
+        myOpMode = opmode;
+        init();
+    }
+
+    public AprilTagPoseEstimator(HardwareMap hardwareMap) {
+        myHardwareMap = hardwareMap;
+        init();
+    }
```

Comment:
Sorry, I missed this in our code review previously. Having two different constructors is odd and confusing. And generally, for proper componentization, a really big object like "LinearOpMode" shouldn't be passed into a class. Rather, the specific objects from "LinearOpMode" that the class needs should be passed in as parameters on the constructor. In this case it's the Telemetry and hardwareMap. Passing the needed objects via the constructor makes the dependencies of that class much clearer.

I know that the sample code passes in "LinearOpMode" in that fashion, but that's bad style.